command/agent/config_parse.go

```diff
@@ -1014,6 +1014,11 @@ func parsePlugins(result *[]*config.PluginConfig, list *ast.ObjectList) error {
 		// Get the current plugin object
 		listVal := list.Items[i]
 
+		// Deal with json->hcl AST parsing incorrectness when directly nested
+		// items show up as additional keys. This currently only affects plugin
+		// configuration because args is not necessary. All other fields in the config
+		// have multiple keys and parse from json into the AST correctly.
```

Comment:
The comment is right in that we only want to target this one - but I don't agree with the rationale.

While it's true that other config stanzas require multiple entries, a user may still set one field only triggering this condition and without `unwrapLegacyHCLObjectKeysFromJSON` call, the error message might not be comprehensible.  However, given this is only relevant for nested entries, I don't see it happening elsewhere.

Beyond 0.9, we want to increase our test coverage.  I noticed that other `parse*` functions have `list = list.Elem()` call that's missing here, correctly so but not sure why.  Also, I noticed that it's easy to get parser to crash with simple config files, like:

```
# basic client stanza
{"client": {"server_join": {"retry_max": 3}}}

# sentinel path that seems similar to this one
{"sentinel": {
		  "import": {"custom-plugin": {"path": "/custom-path"}}
}}
```

In both of these cases, `list = list.Elem()` causes us to get an empty list that causes a crash when indexed.  The pattern causing them affect many stanzes here.

Reply:
agreed, there are more lingering bugs in json parsing. Found this open ticket from 2016 :( https://github.com/hashicorp/nomad/issues/1290. Let's track this for the next minor release, +1 for more test cases

Reply:
Fixed the wording of the comment